### ISSUE TITLE
TypedDict inferance return "Any" or "Object" where it can infer the types.

### ISSUE THREAD

**Issue by fcolecumberri** [microsoft/pyright#9903]
**Describe the bug**

```python

from typing import TypedDict


class MyDict(TypedDict):
    a: int
    b: str


run_time_str = input("")

instance: MyDict = {
    "a": 0,
    "b": "0",
}

val_1 = instance.get(run_time_str, "")

# At this point val_1 is reported as "Any | str"
# However pyright should be able to infer val_1 as "int | str"
# I mean, what else could it be?

val_2 = instance[run_time_str]

# At this point val_2 is reported as "Any | str"
# If run_time_str is not in instance, it will raise without setting,
# so val_2 should also be (Maybe here I'm wrong) "int | str"

assert isinstance(val_1, int) or isinstance(val_1, str)  # Notice that this is doing nothing, val_1 can only be int or str

val_f = val_1

# Now val_f is reported as "int | str" since the assert forced pyright to have a correct inference


# Now, for some reason, types are better inferred with normal typed dicts (A.K.A. non TypedDict)

instance_2: dict[str, int|str] = {
    "a": 0,
    "b": "0",
}

val_3 = instance_2.get(run_time_str, "")  # now val_3 is "int | str"
val_4 = instance_2[run_time_str]  # now val_4 is "int | str"


```

This is specially problematic in a case like this:

```python

from typing import TypedDict


class MyDict(TypedDict):
    a: int
    b: int

instance: MyDict = {
    "a": 0,
    "b": 0,
}

sample_int = int(0)

for k, v in instance.items():  # For some reason, v is "object"
    sample_int.to_bytes()  # This is OK
    v.to_bytes()  # error: Cannot access attribute "to_bytes" for class "object"

for k, v in instance.items():
    assert isinstance(v, int)  # Doing nothing since v can only be int
    v.to_bytes()  # Now it works


instance_2 = {  # Here I am not even type-hinting and yet works better than a TypedDict
    "a": 0,
    "b": 0,
}

for k, v in instance_2.items():
    v.to_bytes()  # Here it works

```

The problem yields on the idea that typing with TypedDict somehow renders a worst inference results than a less restrictive type-hint or not type-hinting at all.

**VS Code extension or command-line**
I have used VS Extension `v1.1.394`, also tested on another machine via CLI with `pyright 1.1.391` and tested on pyright-play.net all of those generates the same result.

**Comment by erictraut** [comment]
Pyright is working correctly here and in conformance with the typing spec, so I don't consider this a bug.

A `TypedDict` is a structural type, and it doesn't describe all possible items in a compatible dictionary. That means additional items (that are not described by the `TypedDict` definition) may be present. When calling the method `get` with an unknown (variable) key, a type checker must assume that you're accessing one of those other items.

There is an in-progress [PEP 728](https://peps.python.org/pep-0728/) that introduces the concept of "closed" TypedDicts, but it is not yet complete. Pyright has experimental support for it if you'd like to play with it.

Also, it's important to note that no type inference is occurring with the `get` method call. Type checkers handle TypedDict methods by synthesizing overloads for the class. In this case, the following overloads are synthesized for the `get` method:

```python
class MyDict:
    ...
    @overload
    def get(self, k: Literal['a']) -> int: ...
    @overload
    def get(self, k: Literal['b']) -> str: ...
    @overload
    def get(self, k: str) -> Any: ...
```

The final overload is provided as a fallback for cases where the provided argument is not a literal.

So there isn't any type inference involved here. Pyright is simply applying standard typing rules for evaluating a call to an overloaded function.
